README.md
# PicoMath

PicoMath is a header-only C++ library for math expression evaluation. PicoMath is very fast, simple, and customizable.

PicoMath was created to evaluate CSS-like expressions, like `"100% - 20px"`, in the UI layout engine of my 2D game engine, Shinobit Engine. The main design goal of the library is to evaluate the expressions one single time as fast as possible, as layout usually happens when a scene is loaded, or a new UI component is shown, but they are not usually evaluated every frame.
Another goal was to make the library as simple as possible, the engine is compiled to WASM and size of the binary matters.

This library is currently used in the MMO browser game [GoBattle.io](http://gobattle.io)

![Build status](https://github.com/Nitrillo/picomath/actions/workflows/ci.yml/badge.svg)

## Features
* Header only C++ library: easy integration in your project
* Very simple and fast
* Zero allocations in the hot path
    * Uses C++17 `std::string_view` to lookup variables without copying strings
* Single pass: The evaluation is performed while parsing.
* Built-in math functions (e.g. `cos(pi)`)
* Custom units and percentages (e.g `100% - 10px`)
* User defined variables (e.g. `(x + y) * (x + y)`)
* User defined functions with multiple arguments (e.g. `avg(10, 20, 30)`)
* Uses standard C++ containers

## How to integrate in your project

Just copy the file `/include/picomath.hpp` in your project

## Usage

Simple evaluation:
```cpp
#include <picomath.hpp>

using namespace picomath;

PicoMath pm;

auto result = pm.evalExpression("sqrt(100 - 20)");
if (result.isOk()) {
    double r = result.getResult();
    ...
}

```

Using variables:
```cpp
#include <picomath.hpp>

using namespace picomath;

PicoMath pm;
auto &x = pm.addVariable("x");

x = 0.0;
while (x < 100.0) {
    // Same expression evaluated with different values of `x`
    auto result = pm.evalExpression("x * x * x");
    if (result.isOk()) {
        double r = result.getResult();
        ...
    }
    x += 1.0;
}

```

Using units:
```cpp
#include <picomath.hpp>

using namespace picomath;

PicoMath pm;
pm.addUnit("km") = 1000.0;
pm.addUnit("cm") = 0.01;

auto result = pm.evalExpression("0.5km + 20cm");
if (result.isOk()) {
    double r = result.getResult();
    ...
}

```



## Test

```shell
# build test binaries
make

# run tests
make test

# run bench tests
make bench
```

The default test binaries will be built in release mode. You can make Debug test binaries as well:

```shell
make clean
make debug
make test
```


## Benchmarks

PicoMath is pretty fast as it evaluates the expression without allocating memory.
Compared to a basic strtolower of the string `(2 + 2) * 4 / 10 - 20.02`, using the method
```
std::transform(data.begin(), data.end(), data.begin(), [](unsigned char c) { return std::tolower(c); }));
```
the evaluation of the same string takes only around 25% more time.
Please run the benchmark in your machine to validate the results.

```
Run on (12 X 2900 MHz CPU s)
CPU Caches:
  L1 Data 32K (x6)  
  L1 Instruction 32K (x6)
  L2 Unified 262K (x6)
  L3 Unified 12582K (x1)
----------------------------------------------------------
Benchmark                   Time           CPU Iterations
----------------------------------------------------------
Baseline tolower           43 ns         43 ns   16254876
Simple expression          54 ns         54 ns   12960515
```

PicoMath was designed to evaluate a expression in one pass, so to reevaluate the same expression requires parsing again the expression.

Some other libraries use bytecode or ASTs to improve the runtime cost of multiple evaluations of the same expression, but that increases complexity, size and number of allocations.
In my tests comparing PicoMath to TinyExpr, you need to evaluate at least three times the same expression to get some benefit of the extra steps.  


# Contributing and License

Contributors are welcome! :sparkles: please see [CONTRIBUTING](CONTRIBUTING.md) for more info.

PicoMath is licensed under a `BSD 3-Clause` license, see [LICENSE](LICENSE.md) for more info.

Please let me know if you use the library in other projects and I can add the link in the readme.
##  Attribution

This repository was created using the HPP-SKEL repository: https://github.com/mapbox/hpp-skel




CHANGELOG.md
## May 27, 2022

* PicoMath is born!


test/test.cpp
#include <picomath.hpp>
#define CATCH_CONFIG_MAIN
#include <catch.hpp>

using namespace picomath; // NOLINT

auto AreSame(number_t a, number_t b) -> bool {
    return fabs(a - b) < FLT_EPSILON;
}

TEST_CASE("Simple expressions") {
    PicoMath ctx;
    REQUIRE(AreSame(ctx.evalExpression("2 + 2").getResult(), 4.0));
    REQUIRE(AreSame(ctx.evalExpression("(2 + 2) * 4").getResult(), 16.0));
    REQUIRE(AreSame(ctx.evalExpression("4 * (2 + 2)").getResult(), 16.0));
    REQUIRE(AreSame(ctx.evalExpression("100 / 10").getResult(), 10.0));
    REQUIRE(AreSame(ctx.evalExpression("100 / (5 + 5)").getResult(), 10.0));
    REQUIRE(AreSame(ctx.evalExpression("(10 * 10) / (5 + 5)").getResult(), 10.0));
    REQUIRE(AreSame(ctx.evalExpression("2 + 2 * 4").getResult(), 10.0));
    REQUIRE(AreSame(ctx.evalExpression("2 + 2 * -4").getResult(), -6.0));
    REQUIRE(AreSame(ctx.evalExpression("2.003 * 1000").getResult(), 2003.0));
    REQUIRE(AreSame(ctx.evalExpression(".1234").getResult(), 0.1234));
    REQUIRE(AreSame(ctx.evalExpression("123.1234").getResult(), 123.1234));
    REQUIRE(AreSame(ctx.evalExpression("-(2+2)").getResult(), -4));
    REQUIRE(ctx.evalExpression("(2+2)").isOk());
    REQUIRE_FALSE(ctx.evalExpression("(2+2)").isError());
}

TEST_CASE("Built-in functions") {
    PicoMath ctx;
    REQUIRE(AreSame(ctx.evalExpression("cos(10) - cos(10) + 1").getResult(), 1.0));
    REQUIRE(AreSame(ctx.evalExpression("min(4, 3, 2, 1)").getResult(), 1.0));
    REQUIRE(AreSame(ctx.evalExpression("max(4, 3, 2, 1)").getResult(), 4.0));
    REQUIRE(AreSame(ctx.evalExpression("(sqrt(10000) * 4) / 100.00").getResult(), 4.0));
    REQUIRE(AreSame(ctx.evalExpression("max(cos(10) - cos(10) + 1, sqrt(10000))").getResult(), 100.0));
}

TEST_CASE("Custom functions") {
    PicoMath ctx;
    ctx.addFunction("multiply", [](size_t argc, const argument_list_t &args) -> Result {
        if (argc != 2) {
            return {"Invalid number of arguments"};
        }
        return {args[0] * args[1]};
    });

    REQUIRE(AreSame(ctx.evalExpression("multiply(8, 2)").getResult(), 16.0));
    REQUIRE(AreSame(ctx.evalExpression("multiply(2, 8)").getResult(), 16.0));
    REQUIRE(ctx.evalExpression("multiply(8)").isError());
    REQUIRE(ctx.evalExpression("multiply(8, 2, 3)").isError());
}

TEST_CASE("Custom units") {
    PicoMath ctx;
    ctx.addUnit("%") = 500.0 / 100.0;
    REQUIRE(AreSame(ctx.evalExpression("100%").getResult(), 500.0));
    REQUIRE(AreSame(ctx.evalExpression("50%").getResult(), 250.0));
    REQUIRE(AreSame(ctx.evalExpression("-1%").getResult(), -5.0));
    REQUIRE(AreSame(ctx.evalExpression("-200%").getResult(), -1000.0));

    ctx.addUnit("km") = 1000.0;
    REQUIRE(AreSame(ctx.evalExpression("1km").getResult(), 1000.0));
    REQUIRE(AreSame(ctx.evalExpression("100km").getResult(), 100000.0));
    REQUIRE(AreSame(ctx.evalExpression("100km - 10").getResult(), 100000.0 - 10.0));
}

TEST_CASE("Invalid expressions") {
    PicoMath ctx;
    REQUIRE(ctx.evalExpression("").isError());
    REQUIRE(ctx.evalExpression("(").isError());
    REQUIRE(ctx.evalExpression("*").isError());
    REQUIRE(ctx.evalExpression("3*").isError());
    REQUIRE(ctx.evalExpression("3 /").isError());
    REQUIRE(ctx.evalExpression("notfound").isError());
    REQUIRE(ctx.evalExpression("notfound()").isError());
    REQUIRE(ctx.evalExpression("notfound(").isError());
    REQUIRE(ctx.evalExpression("2px").isError());
    REQUIRE(ctx.evalExpression("cos(").isError());
    REQUIRE(ctx.evalExpression("cos(1,").isError());
    REQUIRE(ctx.evalExpression("((((((((()))").isError());
    REQUIRE(ctx.evalExpression("{").isError());
    REQUIRE(ctx.evalExpression("-23px").isError());
    REQUIRE(ctx.evalExpression("cos(1,2,3,4,5,6,7,8,9,10)").isError());
    REQUIRE(ctx.evalExpression("-(2*").isError());
}

TEST_CASE("Multi-expressions") {
    PicoMath ctx;
    auto     expression = ctx.evalMultiExpression("-2, 2 * 3, (10 - 5) * 5, sin(0), pi");
    Result   out;
    REQUIRE(expression.evalNext(&out));
    REQUIRE(AreSame(out.getResult(), -2));
    REQUIRE(expression.evalNext(&out));
    REQUIRE(AreSame(out.getResult(), 6));
    REQUIRE(expression.evalNext(&out));
    REQUIRE(AreSame(out.getResult(), 25));
    REQUIRE(expression.evalNext(&out));
    REQUIRE(AreSame(out.getResult(), 0));
    REQUIRE(expression.evalNext(&out));
    REQUIRE(AreSame(out.getResult(), M_PI));
}

TEST_CASE("Variables") {
    PicoMath ctx;
    auto &   x = ctx.addVariable("x");
    auto &   y = ctx.addVariable("y");

    x = 1;
    y = 1;
    REQUIRE(AreSame(ctx.evalExpression("x*x + y*y").getResult(), 2.0));

    x = 2;
    y = 1;
    REQUIRE(AreSame(ctx.evalExpression("x*x + y*y").getResult(), 5.0));

    x = 1;
    y = 2;
    REQUIRE(AreSame(ctx.evalExpression("x*x + y*y").getResult(), 5.0));

    x = 0;
    y = 2;
    REQUIRE(AreSame(ctx.evalExpression("x*x + y*y").getResult(), 4.0));
    REQUIRE(AreSame(ctx.evalExpression("max(x,y) * 2").getResult(), 4.0));
    REQUIRE(AreSame(ctx.evalExpression("min(x,y) * 2").getResult(), 0.0));
}


examples/unit.cpp
#include <iostream>
#include <picomath.hpp>

using PM = picomath::PicoMath;

auto main() -> int {
    PM pm;

    // Device pixel ratio: each virtual pixel = 2 pixels in screen
    auto &devicePixelRatio = pm.addUnit("px");
    devicePixelRatio       = 2.0;

    // Container size 350px
    auto &percentage = pm.addUnit("%");
    percentage       = 350.0 * devicePixelRatio / 100.0;

    // Calculate inner width with a 5px padding in each side
    auto result = pm.evalExpression("100% -5px * 2");
    if (result.isOk()) {
        std::cout << "Result: " << result.getResult() << std::endl;
    } else {
        std::cout << result.getError() << std::endl;
    }

    // Calculate x position of 100px element centered in container
    result = pm.evalExpression("50% - 100px/2");
    if (result.isOk()) {
        std::cout << "Result: " << result.getResult() << std::endl;
    } else {
        std::cout << result.getError() << std::endl;
    }
}

scripts/format.sh
#!/usr/bin/env bash

set -eu
set -o pipefail

: '

Runs clang-format on the code in include/

Return `1` if there are files to be formatted, and automatically formats them.

Returns `0` if everything looks properly formatted.

'
 # Set up the environment by installing mason and clang++
./scripts/setup.sh --config local.env
source local.env

# Add clang-format as a dep
mason install clang-format ${MASON_LLVM_RELEASE}
mason link clang-format ${MASON_LLVM_RELEASE}

# Run clang-format on all cpp and hpp files in the /src directory
find include/ bench/ test/ examples/ -type f -name '*.hpp' -or -name '*.cpp' \
 | xargs -I{} clang-format -i -style=file {}

# Print list of modified files
dirty=$(git ls-files --modified include/ bench/ test/)

if [[ $dirty ]]; then
    echo "The following files have been modified:"
    echo $dirty
    exit 1
else
    exit 0
fi

scripts/coverage.sh
#!/usr/bin/env bash

set -eu
set -o pipefail

# http://clang.llvm.org/docs/UsersManual.html#profiling-with-instrumentation
# https://www.bignerdranch.com/blog/weve-got-you-covered/

# automatically setup environment

./scripts/setup.sh --config local.env
source local.env

make clean
export CXXFLAGS="-fprofile-instr-generate -fcoverage-mapping"
export LDFLAGS="-fprofile-instr-generate"
mason install llvm-cov ${MASON_LLVM_RELEASE}
mason link llvm-cov ${MASON_LLVM_RELEASE}
make debug
rm -f *profraw
rm -f *gcov
rm -f *profdata
LLVM_PROFILE_FILE="code-%p.profraw" make test
CXX_MODULE="./cmake-build/unit-tests"
llvm-profdata merge -output=code.profdata code-*.profraw
llvm-cov report ${CXX_MODULE} -instr-profile=code.profdata -use-color
llvm-cov show ${CXX_MODULE} -instr-profile=code.profdata src/*.cpp --path-equivalence -use-color
llvm-cov show ${CXX_MODULE} -instr-profile=code.profdata src/*.cpp --path-equivalence -use-color --format html > /tmp/coverage.html
echo "open /tmp/coverage.html for HTML version of this report"


include/picomath.hpp
/**
 * @file picomath.hpp
 * @author Cesar Guirao Robles (a.k.a. Nitro) <cesar@no2.es>
 * @brief Math expression evaluation. BSD 3-Clause License
 * @version 1.0.0
 * @date 2022-06-01
 *
 * @copyright Copyright (c) 2022, Cesar Guirao Robles (a.k.a. Nitro) <cesar@no2.es>
 *
 */
#ifndef PICOMATH_HPP
#define PICOMATH_HPP

#include <array>
#include <cmath>
#include <limits>
#include <map>
#include <memory>
#include <string>
#include <string_view>

namespace picomath {

#ifndef PM_MAX_ARGUMENTS
#define PM_MAX_ARGUMENTS 8
#endif

#define PM_LIKELY(x)   __builtin_expect((x), 1)
#define PM_UNLIKELY(x) __builtin_expect((x), 0)
#define PM_INLINE      inline __attribute__((always_inline))

class Result;
class Expression;
class PicoMath;

#ifdef PM_USE_FLOAT
using number_t = float;
#else
using number_t = double;
#endif
using argument_list_t        = std::array<number_t, PM_MAX_ARGUMENTS>;
using error_t                = std::string;
using custom_function_many_t = Result (*)(size_t argc, const argument_list_t &list);
using custom_function_1_t    = number_t (*)(number_t);

class Result {
    friend class Expression;

    number_t                 result;
    std::unique_ptr<error_t> error;

  public:
    Result(number_t value = 0) noexcept : result(value), error() { // NOLINT
    }

    Result(std::string &&description) noexcept
        : result(0), error(std::make_unique<error_t>(std::move(description))) { // NOLINT
    }

    [[nodiscard]] auto isError() const -> bool {
        return error != nullptr;
    }

    [[nodiscard]] auto isOk() const -> bool {
        return error == nullptr;
    }

    [[nodiscard]] auto getError() const -> const char * {
        return error->c_str();
    }

    [[nodiscard]] auto getResult() const -> number_t {
        return result;
    }
};

#define PM_FUNCTION_2(fun)                                                                                             \
    [](size_t argc, const argument_list_t &args) -> Result {                                                           \
        if (argc != 2) {                                                                                               \
            return {"Two arguments needed"};                                                                           \
        }                                                                                                              \
        return fun(args[0], args[1]);                                                                                  \
    }

class PicoMath {
    friend Expression;

    struct Function {
        enum Type
        {
            FunctionMany = 0,
            Function1    = 1
        } type;
        custom_function_many_t many;
        custom_function_1_t    f1;
    };

    std::map<std::string, number_t, std::less<>> variables{};
    std::map<std::string, number_t, std::less<>> units{};
    std::map<std::string, Function, std::less<>> functions{};

  public:
    /**
     * @brief Adds a variable to the parsing context.
     * PicoMath returns a reference to entry inside the map of variables, that can be changed
     * during multiple evaluations
     *
     * @param name Name of the variable
     * @return number_t& Reference to variable's value
     */
    auto addVariable(const std::string &name) -> number_t & {
        return variables[name];
    }

    /**
     * @brief Adds a scale unit to the parsing context
     * PicoMath returns a reference to entry inside the map of variables, that can be changed
     * during multiple evaluations.
     *
     * @param name Name of the unit
     * @return number_t& Reference to the unit's scale value
     */
    auto addUnit(const std::string &name) -> number_t & {
        return units[name];
    }

    /**
     * @brief Adds a custom function to the parsing context
     * This overload allows the user to pass a function that can handle multiple arguments.
     *
     * @param name Name of the function
     * @param func Function pointer
     */
    auto addFunction(const std::string &name, custom_function_many_t func) -> void {
        auto &function = functions[name];
        function.type  = Function::Type::FunctionMany;
        function.many  = func;
    }

    /**
     * @brief Adds a custom function to the parsing context
     *
     * @param name Name of the function
     * @param func Function pointer
     */
    auto addFunction(const std::string &name, custom_function_1_t func) -> void {
        auto &function = functions[name];
        function.type  = Function::Type::Function1;
        function.f1    = func;
    }

    PicoMath() {
        // Constants
        addVariable("pi") = static_cast<number_t>(M_PI);
        addVariable("e")  = static_cast<number_t>(M_E);

        // Built-in functions
        addFunction("abs", std::abs);
        addFunction("ceil", std::ceil);
        addFunction("floor", std::floor);
        addFunction("round", std::round);
        addFunction("ln", std::log);
        addFunction("log", std::log10);
        addFunction("cos", std::cos);
        addFunction("sin", std::sin);
        addFunction("acos", std::acos);
        addFunction("asin", std::asin);
        addFunction("cosh", std::cosh);
        addFunction("sinh", std::sinh);
        addFunction("tan", std::tan);
        addFunction("tanh", std::tanh);
        addFunction("sqrt", std::sqrt);
        addFunction("atan2", PM_FUNCTION_2(std::atan2));
        addFunction("pow", PM_FUNCTION_2(std::pow));
        addFunction("min", [](size_t argc, const argument_list_t &args) -> Result {
            number_t result{std::numeric_limits<number_t>::max()};
            size_t   i = 0;
            while (i < argc) {
                result = std::min(args[i], result);
                i++;
            }
            return result;
        });
        addFunction("max", [](size_t argc, const argument_list_t &args) -> Result {
            number_t result{std::numeric_limits<number_t>::min()};
            size_t   i = 0;
            while (i < argc) {
                result = std::max(args[i], result);
                i++;
            }
            return result;
        });
    }

    /**
     * @brief Evaluates the expression and returns a value or an error if the expression is invalid
     *
     * @param expression Expression to evaluate
     * @return Result Result containing the result value or an error
     */
    auto evalExpression(const char *expression) -> Result;

    /**
     * @brief Creates a Expression that can be used to evaluate multiple expression separated by commas.
     *
     * @param expression MultiExpression to evaluate
     * @return Expression Expression object maintaining the context of the multi-evaluation
     */
    auto evalMultiExpression(const char *expression) -> Expression;
};

class Expression {
    friend PicoMath;

    const PicoMath &context;
    const char *    originalStr{};
    const char *    str{};

    Expression(const PicoMath &picomathContext, const char *expression)
        : context(picomathContext), originalStr(expression), str(expression) {
    }

    auto evalSingle() -> Result {
        Result ret = evalExpression();
        consumeSpace();
        if (PM_LIKELY(isEOF())) {
            return ret;
        }
        return generateError("Invalid characters after expression");
    }

  public:
    /**
     * @brief Evaluates the next expression in the multi expression context
     *
     * @param outResult Result of the evaluation
     * @return true The evaluation succeeded and the outResult parameter contains a valid output value.
     * @return false There are no more expressions to evaluate
     */
    auto evalNext(Result *outResult) -> bool {
        consumeSpace();
        if (isEOF()) {
            return false;
        }
        *outResult = evalExpression();
        consumeSpace();
        if (peek() == ',') {
            consume();
            consumeSpace();
        }
        return true;
    }

  private:
    auto generateError(const char *error, std::string_view identifier = {}) const -> Result {
        std::string out = "In character " + std::to_string(static_cast<int>(str - originalStr - 1)) + ": ";
        out += error;
        if (identifier.empty()) {
            out += " found: ";
            out += isEOF() ? "End of string" : str;
        } else {
            out += ' ';
            out += '`';
            out += identifier;
            out += '`';
        }
        return out;
    }

    PM_INLINE auto evalExpression() -> Result {
        consumeSpace();
        if (PM_UNLIKELY(isEOF())) {
            return generateError("Unexpected end of the string");
        }
        return parseAddition();
    }

    PM_INLINE auto consume() -> char {
        return *str++;
    }

    [[nodiscard]] PM_INLINE auto peek() const -> char {
        return *str;
    }

    PM_INLINE void consumeSpace() {
        while (*str == ' ' || *str == '\t' || *str == '\r' || *str == '\n') {
            consume();
        }
    }

    [[nodiscard]] PM_INLINE auto isDigit() const -> bool {
        return *str >= '0' && *str <= '9';
    }

    [[nodiscard]] PM_INLINE auto isAlpha() const -> bool {
        return (*str >= 'a' && *str <= 'z') || (*str >= 'A' && *str <= 'Z') || (*str == '_');
    }

    [[nodiscard]] PM_INLINE auto isUnitChar() const -> bool {
        return isAlpha() || *str == '%';
    }

    [[nodiscard]] PM_INLINE auto isEOF() const -> bool {
        return *str == 0;
    }

    PM_INLINE auto parseFunction(std::string_view identifier) noexcept -> Result {
        auto f = context.functions.find(identifier);
        if (PM_UNLIKELY(f == context.functions.end())) {
            return generateError("Unknown function", identifier);
        }

        // Consume '('
        consume();
        consumeSpace();

        std::array<number_t, PM_MAX_ARGUMENTS> arguments{};
        size_t                                 argc = 0;

        if (peek() != ')') {
            while (true) {
                if (PM_UNLIKELY(argc == PM_MAX_ARGUMENTS)) {
                    return generateError("Too many arguments");
                }
                Result argument = evalExpression();
                if (PM_UNLIKELY(argument.isError())) {
                    return argument;
                }
                arguments[argc] = argument.result;
                argc++;
                consumeSpace();
                if (peek() != ',') {
                    break;
                }
                consume();
            }
        }

        if (PM_UNLIKELY(peek() != ')')) {
            return generateError("Expected ')'");
        }
        consume();

        if (f->second.type == PicoMath::Function::Type::Function1) {
            if (PM_UNLIKELY(argc != 1)) {
                return generateError("One argument required");
            }
            return {f->second.f1(arguments[0])};
        }
        Result ret = f->second.many(argc, arguments);
        if (PM_UNLIKELY(ret.isError())) {
            // Improve information in errors generated inside functions
            return generateError(ret.getError(), identifier);
        }
        return ret;
    }

    PM_INLINE auto parseParenthesized() -> Result {
        consume();
        consumeSpace();
        Result exp = evalExpression();
        if (PM_UNLIKELY(exp.isError())) {
            return exp;
        }
        consumeSpace();
        // consume ')'
        if (PM_UNLIKELY(peek() != ')')) {
            return generateError("Expected ')'");
        }
        consume();
        return exp;
    }

    auto parsePrefixUnaryOperator() -> Result {
        char op = consume();
        consumeSpace();
        Result unary = parseSubExpression();
        if (PM_UNLIKELY(unary.isError())) {
            return unary;
        }
        if (op == '-') {
            unary.result = -unary.result;
        }
        return unary;
    }

    PM_INLINE auto parseVariableOrFunction() noexcept -> Result {
        const char *start = str;
        size_t      size  = 0;
        do {
            consume();
            size++;
        } while (isAlpha());

        std::string_view identifier{start, size};
        consumeSpace();
        if (peek() == '(') {
            // function call
            return parseFunction(identifier);
        }
        auto f = context.variables.find(identifier);
        if (PM_UNLIKELY(f == context.variables.end())) {
            return generateError("Unknown variable", identifier);
        }
        return {f->second};
    }

    PM_INLINE auto parseNumber() noexcept -> Result {
        number_t ret = 0;

        while (isDigit()) {
            ret *= 10;
            ret += *str - '0';
            consume();
        }
        // Decimal point
        if (peek() == '.') {
            consume();
            number_t weight = 1;
            while (isDigit()) {
                weight /= 10;
                ret += (*str - '0') * weight;
                consume();
            }
        }

        // Space before units
        consumeSpace();

        // Units and percentage
        if (isUnitChar()) {
            const char *start = str;
            size_t      size  = 0;
            do {
                consume();
                size++;
            } while (isUnitChar());

            std::string_view identifier{start, size};

            auto f = context.units.find(identifier);
            if (PM_UNLIKELY(f == context.units.end())) {
                return generateError("Unknown unit", identifier);
            }
            ret = ret * f->second;
        }
        return ret;
    }

    PM_INLINE auto parseSubExpression() -> Result {
        if (isDigit() || peek() == '.') {
            // Number
            return parseNumber();
        }
        if (peek() == '(') {
            // Parenthesized expression
            return parseParenthesized();
        }
        if (peek() == '-' || peek() == '+') {
            // Prefix unary operator
            return parsePrefixUnaryOperator();
        }
        if (isAlpha()) {
            // Variable or function
            return parseVariableOrFunction();
        }
        return generateError("Invalid character");
    }

    auto parseAddition() -> Result {
        consumeSpace();
        Result left = parseMultiplication();
        if (PM_UNLIKELY(left.isError())) {
            return left;
        }
        consumeSpace();
        while (*str == '+' || *str == '-') {
            char op = consume();
            consumeSpace();
            Result right = parseMultiplication();
            if (PM_UNLIKELY(right.isError())) {
                return right;
            }
            if (op == '+') {
                left.result += right.result;
            } else {
                left.result -= right.result;
            }
            consumeSpace();
        }
        return left;
    }

    auto parseMultiplication() -> Result {
        consumeSpace();
        Result left = parseSubExpression();
        if (PM_UNLIKELY(left.isError())) {
            return left;
        }
        consumeSpace();
        while (*str == '*' || *str == '/') {
            char op = consume();
            consumeSpace();
            Result right = parseSubExpression();
            if (PM_UNLIKELY(right.isError())) {
                return right;
            }
            if (op == '*') {
                left.result *= right.result;
            } else if (op == '/') {
                left.result /= right.result;
            }
            consumeSpace();
        }
        return left;
    }
};

inline auto PicoMath::evalExpression(const char *expression) -> Result {
    Expression exp(*this, expression);
    return exp.evalSingle();
}

inline auto PicoMath::evalMultiExpression(const char *expression) -> Expression {
    return {*this, expression};
}

} // namespace picomath
#endif


examples/function.cpp
#include <iostream>
#include <picomath.hpp>

using PM = picomath::PicoMath;

auto main() -> int {
    PM pm;

    pm.addFunction("avg", [](size_t argc, const picomath::argument_list_t &args) -> picomath::Result {
        picomath::number_t result{0};
        size_t             i = 0;
        while (i < argc) {
            result += args[i];
            i++;
        }
        return result / static_cast<picomath::number_t>(argc);
    });

    auto result = pm.evalExpression("avg(20, 40, -23, pow(1.4, 5.3), 7, cos(pi))");
    if (result.isOk()) {
        std::cout << "Result: " << result.getResult() << std::endl;
    } else {
        std::cout << result.getError() << std::endl;
    }

    result = pm.evalExpression("avg(50, 100)");
    if (result.isOk()) {
        std::cout << "Result: " << result.getResult() << std::endl;
    } else {
        std::cout << result.getError() << std::endl;
    }

    pm.addFunction("rand", [](picomath::number_t input) -> picomath::number_t { return input * rand() / RAND_MAX; });
    result = pm.evalExpression("rand(100)");
    if (result.isOk()) {
        std::cout << "Result: " << result.getResult() << std::endl;
    } else {
        std::cout << result.getError() << std::endl;
    }
}

bench/run.cpp
#include <algorithm>
#include <benchmark/benchmark.h>
#include <cctype>
#include <picomath.hpp>
#include <string>

static void BM_simpleExpression(benchmark::State &state) // NOLINT google-runtime-references
{
    picomath::PicoMath ctx;
    while (state.KeepRunning()) {
        benchmark::DoNotOptimize(ctx.evalExpression("(2 + 2) * 4 / 10 - 20.02"));
        benchmark::ClobberMemory();
    }
}

static void BM_baseLine(benchmark::State &state) // NOLINT google-runtime-references
{
    std::string data = "(2 + 2) * 4 / 10 - 20.02";
    while (state.KeepRunning()) {
        benchmark::DoNotOptimize(
            std::transform(data.begin(), data.end(), data.begin(), [](unsigned char c) { return std::tolower(c); }));
        benchmark::ClobberMemory();
    }
}

static void BM_functions(benchmark::State &state) // NOLINT google-runtime-references
{
    picomath::PicoMath ctx;
    while (state.KeepRunning()) {
        benchmark::DoNotOptimize(ctx.evalExpression("sqrt(cos(pi) + ceil(sin(pi)))"));
        benchmark::ClobberMemory();
    }
}

static void BM_customUnit(benchmark::State &state) // NOLINT google-runtime-references
{
    picomath::PicoMath ctx;
    ctx.addUnit("px") = 2.0;
    while (state.KeepRunning()) {
        benchmark::DoNotOptimize(ctx.evalExpression("100px * 2"));
        benchmark::ClobberMemory();
    }
}

static void BM_complexExpression(benchmark::State &state) // NOLINT google-runtime-references
{
    picomath::PicoMath ctx;
    ctx.addVariable("x") = 2.0;
    ctx.addVariable("y") = 3.0;
    ctx.addVariable("z") = 5.0;
    ctx.addVariable("w") = 7.0;
    while (state.KeepRunning()) {
        benchmark::DoNotOptimize(ctx.evalExpression("((((x-(y/(z*w)))/(((x-y)*z)-w))/((((x+y)*z)-w)-((x+y)-(z*w))))/"
                                                    "(((((x-y)*z)-w)*((x+y)-(z/w)))*(((x+y)-(z*w))+((x/y)+(z+w)))))"));
        benchmark::ClobberMemory();
    }
}

static void BM_multiExpression(benchmark::State &state) // NOLINT google-runtime-references
{
    picomath::PicoMath ctx;
    while (state.KeepRunning()) {
        auto             expression = ctx.evalMultiExpression("2 * 4, 2, (100 * 10), pow(2,8)");
        picomath::Result result;
        while (expression.evalNext(&result)) {
            if (result.isOk()) {
                benchmark::DoNotOptimize(result.getResult());
            }
        }
        benchmark::ClobberMemory();
    }
}

auto main(int argc, char *argv[]) -> int {
    benchmark::RegisterBenchmark("Baseline tolower", // NOLINT clang-analyzer-cplusplus.NewDeleteLeaks
                                 BM_baseLine);
    benchmark::RegisterBenchmark("Simple expression", // NOLINT clang-analyzer-cplusplus.NewDeleteLeaks
                                 BM_simpleExpression);
    benchmark::RegisterBenchmark("Built-in functions", // NOLINT clang-analyzer-cplusplus.NewDeleteLeaks
                                 BM_functions);
    benchmark::RegisterBenchmark("Custom units", // NOLINT clang-analyzer-cplusplus.NewDeleteLeaks
                                 BM_customUnit);
    benchmark::RegisterBenchmark("Complex expression", // NOLINT clang-analyzer-cplusplus.NewDeleteLeaks
                                 BM_complexExpression);
    benchmark::RegisterBenchmark("Multiexpression", // NOLINT clang-analyzer-cplusplus.NewDeleteLeaks
                                 BM_multiExpression);

    benchmark::Initialize(&argc, argv);
    benchmark::RunSpecifiedBenchmarks();

    return 0;
}